### PR TITLE
Add initial code for chain id support

### PR DESCRIPTION
**Pull request by sydhds**

### CODE REVIEW

File: packages/massa-web3/src/web3/ClientFactory.ts [hunk 1]
```diff
@@ -125,13 +129,15 @@ export class ClientFactory {
...
    * @param provider - Provider from wallet provider to be used by the client.
+   * @param chainId - Chain id matching the network used by the provider
    * @param baseAccount - Base account from the wallet provider to be used by the client.
    * @param retryStrategyOn - Whether to retry failed requests.
    *
    * @returns A promise that resolves to a Client object.
    */
   public static async fromWalletProvider(
     provider: IProviderWalletProvider,
+    chainId: bigint,
```
Comment (Thykof) [inline packages/massa-web3/src/web3/ClientFactory.ts:140]: unused parameter

File: packages/massa-web3/src/web3/accounts/Web3Account.ts [hunk 2]
```diff
@@ -134,9 +141,18 @@ export class Web3Account extends BaseClient implements IBaseAccount {
...
+    view.setBigUint64(0, this.chainId, false);
+
     // sign payload
     const signature: ISignature = await this.sign(
-      Buffer.concat([getBytesPublicKey(this.account.publicKey), bytesCompact]),
+      Buffer.concat([
+        Buffer.from(chainIdBuffer),
+        getBytesPublicKey(this.account.publicKey),
+        bytesCompact,
+      ]),
```
Comment (Thykof) [inline packages/massa-web3/src/web3/accounts/Web3Account.ts:155]: we are duplicating a lot of code here, better write a function that prepare the buffer we want to sign.

File: packages/massa-web3/test/web3/mockData.ts [hunk 3]
```diff
@@ -88,6 +88,7 @@ export const mockNodeStatusInfo: INodeStatus = {
     pos_lock_cycles: 64,
     pos_lookback_cycles: 64,
   },
+  chain_id: BigInt(77658366)
```
Comment (Thykof) [inline packages/massa-web3/test/web3/mockData.ts:91]: 77658366 this constant in tests and mock is duplicated, better declare it somewhere: `export const BUILDNET_CHAIN_ID = 77658366;`

**Review by Thykof — CHANGES_REQUESTED** [review]

**Comment by Thykof** [comment]
I pushed a commit to format the code by doing `npm run fmt`

**Comment by Thykof** [comment]
need to update this: https://github.com/massalabs/massa-web3/blob/08affbe472232e44dcafc19833a70f1ff0a1e3ed/packages/massa-web3/examples/wallet/index.ts#L60

File: packages/massa-web3/src/web3/accounts/Web3Account.ts (hunk 2 above)
Comment (sydhds) [reply packages/massa-web3/src/web3/accounts/Web3Account.ts]: Fixed

File: packages/massa-web3/test/web3/mockData.ts (hunk 3 above)
Comment (sydhds) [reply packages/massa-web3/test/web3/mockData.ts]: Done

**Review by Thykof — CHANGES_REQUESTED** [review]
export the table "network name"=>chain id 

https://github.com/massalabs/station-massa-wallet/pull/816#discussion_r1437046132

File: packages/massa-web3/src/web3/accounts/Web3Account.ts [hunk 4]
```diff
@@ -15,13 +15,104 @@ import { ICallData } from '../../interfaces/ICallData';
...
+
+  /**
+   * Executes a blockchain operation
+   *
+   * @param txData - The transaction data for the operation.
+   * @param operationType - The type of operation to be executed.
+   * @param useRetry - Determines whether to use retry logic in case of failures.
+   * @param errorMessage - Custom error message to throw if operation fails.
+   * @param preExecutionCallback - An optional callback function to be executed before the operation, for any pre-execution logic or validation.
+   * @returns {Promise<string>} - Returns a promise that resolves to the operation ID.
```
Comment (Thykof) [inline packages/massa-web3/src/web3/accounts/Web3Account.ts:59]:
```suggestion
   * @returns Returns a promise that resolves to the operation ID.
```

File: packages/massa-web3/src/web3/accounts/Web3Account.ts [hunk 5]
```diff
@@ -117,225 +208,55 @@ export class Web3Account extends BaseClient implements IBaseAccount {
...
-    if (!contractData.contractDataBinary) {
-      throw new Error(
-        `Expected non-null contract bytecode, but received null.`,
-      );
-    }
+    const preExecutionLogic = async (data: IContractData) => {
+      // Check if SC data exists
+      if (!data.contractDataBinary) {
+        throw new Error(
+          `Expected non-null contract bytecode, but received null.`,
```
Comment (Thykof) [inline packages/massa-web3/src/web3/accounts/Web3Account.ts:237]:
```suggestion
          'Expected non-null contract bytecode, but received null.',
```

File: packages/massa-web3/src/web3/accounts/Web3Account.ts [hunk 6]
```diff
@@ -117,225 +208,55 @@ export class Web3Account extends BaseClient implements IBaseAccount {
...
-    );
-
-    // sign payload
-    const bytesPublicKey: Uint8Array = getBytesPublicKey(
-      this.account.publicKey,
-    );
-    const signature: ISignature = await this.sign(
-      Buffer.concat([bytesPublicKey, bytesCompact]),
+      this.clientConfig.retryStrategyOn,
+      `Call smart contract operation bad response. No results array in json rpc response. Inspect smart contract`,
```
Comment (Thykof) [inline packages/massa-web3/src/web3/accounts/Web3Account.ts:226]:
```suggestion
      'Call smart contract operation bad response. No results array in json rpc response. Inspect smart contract',
```

File: packages/massa-web3/src/web3/accounts/Web3Account.ts
Comment (Thykof) [inline packages/massa-web3/src/web3/accounts/Web3Account.ts:1]: great refacto!

**Review by Thykof — CHANGES_REQUESTED** [review]

**Review by Thykof — APPROVED** [review]

File: packages/web3-utils/src/constants.ts [hunk 7]
```diff
@@ -23,3 +23,20 @@ export const BASE_ACCOUNT_CREATION_COST = fromMAS(0.001);
...
+export const LABNET = 'LabNet';
+export const SANDBOX = 'Sandbox';
+
+export const CHAIN_ID_TO_NETWORK_NAME = {
+  77658377: MAINNET,
+  77658366: BUILDNET,
+  77658383: SECURENET,
+  77658376: LABNET,
+  77: SANDBOX,
+} as const; // type is inferred as the specific, unchangeable structure
```
Comment (Thykof) [inline packages/web3-utils/src/constants.ts:42]: we forgot to add:

```ts
export type ChainId = keyof typeof CHAIN_ID_TO_NETWORK_NAME;
```

Comment (Ben-Rey) [reply packages/web3-utils/src/constants.ts:42]: You're right thanks, i'll add it

**Pull request merged**